crates/bifrost/src/providers/replicated_loglet/rpc_routers.rs

```diff
@@ -55,15 +72,607 @@ impl LogServersRpc {
 /// Cloning this is cheap and all clones will share the same internal trackers.
 #[derive(Clone)]
 pub struct SequencersRpc {
-    pub append: RpcRouter<Append>,
+    connection_manager: RemoteSequencerConnectionManager,
+    max_inflight_records_in_config: usize,
+    record_permits: Arc<Semaphore>,
 }
 
 impl SequencersRpc {
+    /// Creates a new instance from SequencerRpc
+    ///
     /// Registers all routers into the supplied message router. This ensures that
     /// responses are routed correctly.
     pub fn new(router_builder: &mut MessageRouterBuilder) -> Self {
-        let append = RpcRouter::new(router_builder);
+        let rpc_router = RpcRouter::new(router_builder);
+
+        let max_inflight_records_in_config: usize = Configuration::pinned()
+            .bifrost
+            .replicated_loglet
+            .maximum_inflight_records
+            .into();
+
+        let record_permits = Arc::new(Semaphore::new(max_inflight_records_in_config));
+
+        Self {
+            connection_manager: RemoteSequencerConnectionManager::new(rpc_router),
+            max_inflight_records_in_config,
+            record_permits,
+        }
+    }
+
+    pub async fn append<T: TransportConnect>(
+        &self,
+        loglet: &ReplicatedLoglet<T>,
+        payloads: Arc<[Record]>,
+    ) -> Result<LogletCommit, OperationError> {
+        if payloads.len() > self.max_inflight_records_in_config {
```

Comment:
unfortunately semaphore don't expose total number of permits, but only available permits. This can cause unnecessary overshooting of number of permits